### PR TITLE
Add :picasso-layoutlib-compat module to support Picasso with LayoutLib

### PR DESCRIPTION
**Pull request by gamepro65**
Adds support for ComposeUiTooling's @Preview and Paparazzi unit tests by ensuring we execute our Handler commands properly within LayoutLib. Once we move to coroutines this should go away but in the mean time this works as a nice stop gap.

Setting different size/color Bitmaps to represent images in this @Preview demo to keep the example simple

https://user-images.githubusercontent.com/672316/196202982-0b9753f0-9936-4419-a903-2de2fe4fe506.mp4

### CODE REVIEW

File: picasso-layoutlib-compat/src/main/java/com/squareup/picasso3/LayoutLibThreadFactory.kt [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  init {
+    if (layoutLibEnvironment) {
+      // Handler_Delegate sendMessageAtTime drops messages that
+      // don't contain a callback. So we will continue to check
+      // if the message has a callback and add it to the
+      // handlerMessageQueue like normal else check if it is
+      // Picasso's Action trying to be sent, at which point
+      // we want to forward that message to continue processing
+      Handler_Delegate.setCallback { handler, message, uptimeMillis ->
+        if (message.callback != null) {
```
Comment (gamepro65) [inline picasso-layoutlib-compat/src/main/java/com/squareup/picasso3/LayoutLibThreadFactory.kt:40]: Source: https://cs.android.com/android/platform/superproject/+/master:frameworks/layoutlib/bridge/src/android/os/Handler_Delegate.java;l=44;bpv=1;bpt=0

File: picasso-layoutlib-compat/src/main/java/com/squareup/picasso3/LayoutLibThreadFactory.kt [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+      override fun run() {
+        // This is Picasso's thread, we always want to handle messages posted here
+        if (layoutLibEnvironment)
+          Handler_Delegate.setCallback { handler, message, _ -> handler.handleMessage(message) }
+
+        super.run()
+      }
+    }
+
+  companion object {
```
Comment (afollestad) [inline picasso-layoutlib-compat/src/main/java/com/squareup/picasso3/LayoutLibThreadFactory.kt:65]: Super nit: is there a reason to put `private` on the children instead of on the `companion object`?

**Review by afollestad — APPROVED** [review]

File: picasso-layoutlib-compat/build.gradle [hunk 3]
```diff
@@ -0,0 +1,36 @@
+apply plugin: 'com.android.library'
+apply plugin: 'org.jetbrains.kotlin.android'
+apply plugin: 'app.cash.paparazzi'
+apply plugin: 'com.vanniktech.maven.publish'
```
Comment (jrodbx) [inline picasso-layoutlib-compat/build.gradle:4]: I think this should be a jvm module.  From what I can tell, it's only an android module for the purposes of Paparazzi testing?  Solution: make a new module `picasso-paparazzi-sample` and move the PicassoPaparazziTest there.  Eventually, when Paparazzi supports application modules, we can merge that with the currently-existing sample.

File: build.gradle [hunk 4]
```diff
@@ -37,6 +39,11 @@ buildscript {
     runtime: "androidx.compose.runtime:runtime:${versions.composeUi}",
     foundation: "androidx.compose.foundation:foundation:${versions.composeUi}",
     material: "androidx.compose.material:material:${versions.composeUi}",
+    composeUiToolingPreview: "androidx.compose.ui:ui-tooling-preview:${versions.composeUi}",
+    uiTooling: "androidx.compose.ui:ui-tooling:${versions.composeUi}",
+    paparazzi: "app.cash.paparazzi:paparazzi:${versions.paparazzi}",
+    paparazziGradlePlugin: "app.cash.paparazzi:paparazzi-gradle-plugin:${versions.paparazzi}",
```
Comment (jrodbx) [inline build.gradle:45]:
```suggestion
    paparazziPlugin: "app.cash.paparazzi:paparazzi-gradle-plugin:${versions.paparazzi}",
```

File: build.gradle [hunk 5]
```diff
@@ -37,6 +39,11 @@ buildscript {
...
+    layoutlib: "app.cash.paparazzi:layoutlib-native-jdk11:${versions.layoutlib}",
```
Comment (jrodbx) [inline build.gradle:46]: inline version (single use) and use single quotes

File: picasso-layoutlib-compat/api/picasso-layoutlib-compat.api [hunk 6]
```diff
@@ -0,0 +1,9 @@
+public final class com/squareup/picasso3/LayoutLibThreadFactory : java/util/concurrent/ThreadFactory {
+	public static final field Companion Lcom/squareup/picasso3/LayoutLibThreadFactory$Companion;
+	public fun <init> ()V
+	public fun newThread (Ljava/lang/Runnable;)Ljava/lang/Thread;
+}
+
+public final class com/squareup/picasso3/LayoutLibThreadFactory$Companion {
+}
```
Comment (jrodbx) [inline picasso-layoutlib-compat/api/picasso-layoutlib-compat.api:8]: probably shouldn't be in the public API

File: picasso-layoutlib-compat/gradle.properties [hunk 7]
```diff
@@ -0,0 +1,4 @@
+POM_ARTIFACT_ID=picasso-layoutlib-compat
+POM_NAME=Picasso Layoutlib Compat
+POM_DESCRIPTION=Provides a hook for executing ThreadFactory Threads with Layoutlib
```
Comment (jrodbx) [inline picasso-layoutlib-compat/gradle.properties:3]:
```suggestion
POM_DESCRIPTION=Provides a compat ThreadFactory for interop with Layoutlib
```

File: picasso-layoutlib-compat/src/main/java/com/squareup/picasso3/LayoutLibThreadFactory.kt (hunk 2 above)
Comment (jrodbx) [reply picasso-layoutlib-compat/src/main/java/com/squareup/picasso3/LayoutLibThreadFactory.kt:65]: agreed, this should be private

File: picasso-layoutlib-compat/src/test/java/com/squareup/picasso3/PicassoPaparazziTest.kt [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+import com.squareup.picasso3.Picasso.LoadedFrom.MEMORY
+import com.squareup.picasso3.RequestHandler.Result.Bitmap
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(org.junit.runners.BlockJUnit4ClassRunner::class)
+class PicassoPaparazziTest {
+
+  @get:Rule val paparazzi = Paparazzi(snapshotHandler = SnapshotVerifier(0.0))
```
Comment (jrodbx) [inline picasso-layoutlib-compat/src/test/java/com/squareup/picasso3/PicassoPaparazziTest.kt:32]: why add the `SnapshotVerifier(0.0)` vs relying on the default case?

File: picasso-layoutlib-compat/src/test/java/com/squareup/picasso3/PicassoPaparazziTest.kt [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+import android.widget.ImageView.ScaleType.CENTER
+import app.cash.paparazzi.Paparazzi
+import app.cash.paparazzi.SnapshotVerifier
+import com.squareup.picasso3.Picasso.LoadedFrom.MEMORY
+import com.squareup.picasso3.RequestHandler.Result.Bitmap
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(org.junit.runners.BlockJUnit4ClassRunner::class)
```
Comment (jrodbx) [inline picasso-layoutlib-compat/src/test/java/com/squareup/picasso3/PicassoPaparazziTest.kt:29]: not needed?

File: picasso-layoutlib-compat/src/main/java/com/squareup/picasso3/LayoutLibThreadFactory.kt [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+import android.os.Handler_Delegate
+import java.util.concurrent.ThreadFactory
+
+/**
+ * LayoutLib requires assistance bridging the gap to work with ThreadFactory.
+ * This workaround ensures the Handler_Delegate executes all threads that are created
+ * by Picasso allowing RequestHandler's the chance to fulfill a request synchronously.
+ */
+@SuppressLint("SyntheticAccessor")
+class LayoutLibThreadFactory : ThreadFactory {
```
Comment (jrodbx) [inline picasso-layoutlib-compat/src/main/java/com/squareup/picasso3/LayoutLibThreadFactory.kt:28]:
```suggestion
class LayoutlibThreadFactory : ThreadFactory {
```

File: picasso-layoutlib-compat/src/main/java/com/squareup/picasso3/LayoutLibThreadFactory.kt [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+        // to handle messages for Picasso's Threads.
+        if (layoutLibEnvironment)
+          Handler_Delegate.setCallback { handler, message, _ -> handler.handleMessage(message) }
+
+        super.run()
+      }
+    }
+
+  private companion object {
+    private val layoutLibEnvironment by lazy {
```
Comment (jrodbx) [inline picasso-layoutlib-compat/src/main/java/com/squareup/picasso3/LayoutLibThreadFactory.kt:52]:
```suggestion
    private val isLayoutlibPresent by lazy {
```

Comment (jrodbx) [inline picasso-layoutlib-compat/src/main/java/com/squareup/picasso3/LayoutLibThreadFactory.kt:52]: also, let's add a comment explaining why reflection is used for Previews on device, vs an explicit class reference.

File: picasso-layoutlib-compat/src/main/java/com/squareup/picasso3/LayoutLibThreadFactory.kt [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+      // don't contain a callback. So attach a callback which
+      // executes all messages
+      Handler_Delegate.setCallback { handler, message, _ -> handler.handleMessage(message) }
+    }
+  }
+
+  override fun newThread(runnable: Runnable) =
+    object : Thread(runnable) {
+      override fun run() {
+        // Handler_Delegate's callback is ThreadLocal, so we also need
```
Comment (jrodbx) [inline picasso-layoutlib-compat/src/main/java/com/squareup/picasso3/LayoutLibThreadFactory.kt:42]:
```suggestion
        // Handler_Delegate stores thread-local callbacks, so this callback is for Picasso thread message handling
```

File: picasso-layoutlib-compat/src/main/java/com/squareup/picasso3/LayoutLibThreadFactory.kt [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+ * LayoutLib requires assistance bridging the gap to work with ThreadFactory.
+ * This workaround ensures the Handler_Delegate executes all threads that are created
+ * by Picasso allowing RequestHandler's the chance to fulfill a request synchronously.
+ */
+@SuppressLint("SyntheticAccessor")
+class LayoutLibThreadFactory : ThreadFactory {
+
+  init {
+    if (layoutLibEnvironment) {
+      // Handler_Delegate sendMessageAtTime drops messages that
```
Comment (jrodbx) [inline picasso-layoutlib-compat/src/main/java/com/squareup/picasso3/LayoutLibThreadFactory.kt:32]:
```suggestion
      // Otherwise, Handler_Delegate.sendMessageAtTime drops messages without a callback.
```

Also, explain what thread owns this callback and why it's needed?

File: picasso-layoutlib-compat/src/main/java/com/squareup/picasso3/LayoutLibThreadFactory.kt [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+package com.squareup.picasso3
+
+import android.annotation.SuppressLint
+import android.os.Handler_Delegate
+import java.util.concurrent.ThreadFactory
+
+/**
+ * LayoutLib requires assistance bridging the gap to work with ThreadFactory.
+ * This workaround ensures the Handler_Delegate executes all threads that are created
+ * by Picasso allowing RequestHandler's the chance to fulfill a request synchronously.
```
Comment (jrodbx) [inline picasso-layoutlib-compat/src/main/java/com/squareup/picasso3/LayoutLibThreadFactory.kt:25]:
```suggestion
 * by Picasso allowing RequestHandlers the chance to fulfill a request synchronously.
```

File: picasso-paparazzi-sample/src/test/java/com/squareup/picasso3/PicassoPaparazziTest.kt [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+
+        override fun load(picasso: Picasso, request: Request, callback: Callback) {
+          val bitmap = BitmapFactory.decodeStream(javaClass.classLoader!!.getResourceAsStream(request.uri!!.lastPathSegment!!))
+          callback.onSuccess(Bitmap(bitmap, MEMORY))
+        }
+      })
+      .build()
+
+    paparazzi.snapshot(
+      ImageView(paparazzi.context).also {
```
Comment (jrodbx) [inline picasso-paparazzi-sample/src/test/java/com/squareup/picasso3/PicassoPaparazziTest.kt:49]: nit: using apply, would change `it` to `this`.  arguably nicer?

File: picasso-paparazzi-sample/src/test/java/com/squareup/picasso3/PicassoPaparazziTest.kt [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+
+  @Test
+  fun loadsUrlIntoImageView() {
+    val picasso = Picasso.Builder(paparazzi.context)
+      .callFactory { throw AssertionError() } // Removes network
+      .executor(PicassoExecutorService(threadFactory = LayoutLibThreadFactory()))
+      .addRequestHandler(object : RequestHandler() {
+        override fun canHandleRequest(data: Request): Boolean {
+          return data.uri?.lastPathSegment?.run(javaClass.classLoader::getResourceAsStream) != null
+        }
```
Comment (jrodbx) [inline picasso-paparazzi-sample/src/test/java/com/squareup/picasso3/PicassoPaparazziTest.kt:39]: let's use a convention to simplfy this
```suggestion
        override fun canHandleRequest(data: Request): Boolean {
           return "fake" == data.uri!!.scheme
        }
```

File: picasso-paparazzi-sample/src/test/java/com/squareup/picasso3/PicassoPaparazziTest.kt [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+class PicassoPaparazziTest {
+
+  @get:Rule val paparazzi = Paparazzi()
+
+  @Test
+  fun loadsUrlIntoImageView() {
+    val picasso = Picasso.Builder(paparazzi.context)
+      .callFactory { throw AssertionError() } // Removes network
+      .executor(PicassoExecutorService(threadFactory = LayoutLibThreadFactory()))
+      .addRequestHandler(object : RequestHandler() {
```
Comment (jrodbx) [inline picasso-paparazzi-sample/src/test/java/com/squareup/picasso3/PicassoPaparazziTest.kt:36]: let's extract into a `FakeRequestHandler` local class in this same file.

File: picasso-paparazzi-sample/src/test/java/com/squareup/picasso3/PicassoPaparazziTest.kt [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+        override fun canHandleRequest(data: Request): Boolean {
+          return data.uri?.lastPathSegment?.run(javaClass.classLoader::getResourceAsStream) != null
+        }
+
+        override fun load(picasso: Picasso, request: Request, callback: Callback) {
+          val bitmap = BitmapFactory.decodeStream(javaClass.classLoader!!.getResourceAsStream(request.uri!!.lastPathSegment!!))
+          callback.onSuccess(Bitmap(bitmap, MEMORY))
+        }
+      })
```
Comment (jrodbx) [inline picasso-paparazzi-sample/src/test/java/com/squareup/picasso3/PicassoPaparazziTest.kt:45]:
```suggestion
    override fun load(picasso: Picasso, request: Request, callback: Callback) {
      val imagePath = request.uri!!.lastPathSegment!!

      callback.onSuccess(Result.Bitmap(loadBitmap(imagePath)!!, MEMORY))
    }
    
    private fun loadBitmap(imagePath: String): android.graphics.Bitmap? {
      val resourceAsStream = javaClass.classLoader!!.getResourceAsStream(imagePath)
      return BitmapFactory.decodeStream(resourceAsStream)
    }
```

slight cleanup for readability

File: picasso-layoutlib-compat/src/main/java/com/squareup/picasso3/LayoutLibThreadFactory.kt [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+ */
+@SuppressLint("SyntheticAccessor")
```
Comment (jrodbx) [inline picasso-layoutlib-compat/src/main/java/com/squareup/picasso3/LayoutLibThreadFactory.kt:27]: why is this needed?  is it because of the `newThread` implementation as an expression function?

File: build.gradle [hunk 20]
```diff
@@ -37,6 +38,11 @@ buildscript {
     runtime: "androidx.compose.runtime:runtime:${versions.composeUi}",
     foundation: "androidx.compose.foundation:foundation:${versions.composeUi}",
     material: "androidx.compose.material:material:${versions.composeUi}",
+    composeUiToolingPreview: "androidx.compose.ui:ui-tooling-preview:${versions.composeUi}",
+    uiTooling: "androidx.compose.ui:ui-tooling:${versions.composeUi}",
+    paparazzi: "app.cash.paparazzi:paparazzi:${versions.paparazzi}",
+    paparazziPlugin: "app.cash.paparazzi:paparazzi-gradle-plugin:${versions.paparazzi}",
+    layoutlib: 'app.cash.paparazzi:layoutlib-native-jdk11:2022.1.1-canary-f5f9f71',
```
Comment (jrodbx) [inline build.gradle:45]:
```suggestion
    layoutlib: 'app.cash.paparazzi:layoutlib-native-jdk11:2021.2.1-patch1-fa3aa65',
```

this is the latest released/supported version in Paparazzi

File: picasso-layoutlib-compat/api/picasso-layoutlib-compat.api [hunk 21]
```diff
@@ -0,0 +1,5 @@
+public final class com/squareup/picasso3/LayoutLibThreadFactory : java/util/concurrent/ThreadFactory {
```
Comment (jrodbx) [inline picasso-layoutlib-compat/api/picasso-layoutlib-compat.api:1]: Let's chat the module package structure and update the Factory name (lowercased l in Lib)
```suggestion
public final class com/squareup/picasso3/layoutlib/LayoutlibThreadFactory : java/util/concurrent/ThreadFactory {
```

File: picasso-paparazzi-sample/src/main/AndroidManifest.xml [hunk 22]
```diff
@@ -0,0 +1 @@
+<manifest package="com.squareup.picasso3.layoutlibcompat" />
```
Comment (jrodbx) [inline picasso-paparazzi-sample/src/main/AndroidManifest.xml:1]:
```suggestion
<manifest package="com.example.picasso.paparazzi" />
```

File: picasso-layoutlib-compat/src/main/java/com/squareup/picasso3/LayoutLibThreadFactory.kt (hunk 19 above)
Comment (gamepro65) [reply picasso-layoutlib-compat/src/main/java/com/squareup/picasso3/LayoutLibThreadFactory.kt]: Missed refactor

File: picasso-paparazzi-sample/src/test/java/com/example/paparazzi/PicassoPaparazziTest.kt [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+    val picasso = Picasso.Builder(paparazzi.context)
+      .callFactory { throw AssertionError() } // Removes network
+      .executor(PicassoExecutorService(threadFactory = LayoutlibThreadFactory()))
+      .addRequestHandler(FakeRequestHandler())
+      .build()
+
+    paparazzi.snapshot(
+      ImageView(paparazzi.context).apply {
+        scaleType = CENTER
+        picasso.load("fake://cash.app/zkaAooq.jpeg")
```
[the remark this answers is not in the document]
Comment (jrodbx) [reply picasso-paparazzi-sample/src/test/java/com/example/paparazzi/PicassoPaparazziTest.kt:46]: also, rename `src/test/resources/zkaAooq.jpeg` to `src/test/resources/zkaAooq.png`

**Review by jrodbx — COMMENTED** [review]
We'll also need to enable Git LFS for the repo to avoid test resources for `snapshots/images` content.

File: .github/workflows/build.yaml [hunk 24]
```diff
@@ -12,6 +12,11 @@ jobs:
     steps:
       - name: Checkout
         uses: actions/checkout@v3
+        with:
+          lfs: true
+
+      - name: Checkout LFS objects
+        run: git lfs pull
```
Comment (jrodbx) [inline .github/workflows/build.yaml:19]:
```suggestion
```

This should now be covered by the checkout action

File: build.gradle [hunk 25]
```diff
@@ -37,6 +38,11 @@ buildscript {
     runtime: "androidx.compose.runtime:runtime:${versions.composeUi}",
     foundation: "androidx.compose.foundation:foundation:${versions.composeUi}",
     material: "androidx.compose.material:material:${versions.composeUi}",
+    composeUiToolingPreview: "androidx.compose.ui:ui-tooling-preview:${versions.composeUi}",
+    uiTooling: "androidx.compose.ui:ui-tooling:${versions.composeUi}",
+    paparazzi: "app.cash.paparazzi:paparazzi:${versions.paparazzi}",
```
Comment (jrodbx) [inline build.gradle:43]:
```suggestion
```

remove?

File: picasso-sample/build.gradle [hunk 26]
```diff
@@ -45,8 +45,11 @@ dependencies {
   implementation deps.runtime
   implementation deps.foundation
   implementation deps.material
+  implementation deps.uiTooling
+  implementation deps.composeUiToolingPreview
```
Comment (jrodbx) [inline picasso-sample/build.gradle:49]:
```suggestion
```

remove, transitively brought in by uiTooling

File: build.gradle [hunk 27]
```diff
@@ -37,6 +38,11 @@ buildscript {
     runtime: "androidx.compose.runtime:runtime:${versions.composeUi}",
     foundation: "androidx.compose.foundation:foundation:${versions.composeUi}",
     material: "androidx.compose.material:material:${versions.composeUi}",
+    composeUiToolingPreview: "androidx.compose.ui:ui-tooling-preview:${versions.composeUi}",
```
Comment (jrodbx) [inline build.gradle:41]:
```suggestion
```

remove?

File: picasso-layoutlib-compat/src/main/AndroidManifest.xml [hunk 28]
```diff
@@ -0,0 +1 @@
+<manifest package="com.squareup.picasso3.layoutlibcompat" />
```
Comment (jrodbx) [inline picasso-layoutlib-compat/src/main/AndroidManifest.xml:1]: remove, since now a jvm project

**Review by jrodbx — APPROVED** [review]
